### PR TITLE
Add instance status endpoint

### PR DESCRIPTION
**Pull request by tgracchus**
Adding two instance status endpoints in order to check running bakes.
Both are enabled by default

To disable they add to rosco-local.yml
endpoint.status.enable : true

To enable sensitive flag add to rosco-local.yml
endpoint.status.sensitive : false

**/status/all**
```
{  
   "instance":"a74e6995-ab43-40e2-9b40-cc6afe01e048",
   "instances":{  
      "a74e6995-ab43-40e2-9b40-cc6afe01e048":{  
         "status":"RUNNING",
         "bakes":[  
            {  
               "id":"a459c99e-3ee1-41a0-a45d-de48b0fb2e70",
               "state":"RUNNING",
               "result":null,
               "resource_id":"a459c99e-3ee1-41a0-a45d-de48b0fb2e70"
            }
         ]
      },
      "4e51032e-a114-11e6-80f5-76304dec7eb7":{  
         "status":"RUNNING",
         "bakes":[  
            {  
               "id":"6442a7c8-a114-11e6-80f5-76304dec7eb7",
               "state":"RUNNING",
               "result":null,
               "resource_id":"6442a7c8-a114-11e6-80f5-76304dec7eb7"
            }
         ]
      }
   }
}
```

**/status/instance**
It returns a json with the following format
```
{  
   "status":"RUNNING",
   "bakes":[  
      {  
         "id":"d26cb40c-8758-4ff6-a330-c8cd5c8afe94",
         "state":"RUNNING",
         "result":null,
         "resource_id":"d26cb40c-8758-4ff6-a330-c8cd5c8afe94"
      },
      {  
         "id":"89433795-3ddf-4417-a654-e76a47e8e938",
         "state":"RUNNING",
         "result":null,
         "resource_id":"89433795-3ddf-4417-a654-e76a47e8e938"
      }
   ]
}
```

or in the case the instance is not baking anything
```
{  
   "status":"IDLE",
   "bakes":[ ]
}
```

### CODE REVIEW

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    def instanceStatus
+    def instanceIncompleteBakeIds = bakeStore.getThisInstanceIncompleteBakeIds()
+    if (instanceIncompleteBakeIds) {
+      instanceStatus = Status.RUNNING
+      bakes = bakesStatus(instanceIncompleteBakeIds)
+    } else {
+      instanceStatus = Status.IDLE
+      bakes = []
+    }
+    Collections.unmodifiableMap(["status": instanceStatus.name(), "bakes": bakes])
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy]: Please use `return` operator. We've been inconsistent about this in the past, but prefer to use them when it's not a single-line closure or something similar.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/mvc/StatusMvcEndpoint.groovy [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.boot.actuate.endpoint.mvc.EndpointMvcAdapter
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.stereotype.Component
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestMethod
+import org.springframework.web.bind.annotation.ResponseBody
+
+@Component
+class StatusMvcEndpoint extends EndpointMvcAdapter{
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/mvc/StatusMvcEndpoint.groovy]: Missing a space just before the opening brace.

File: rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import com.netflix.spinnaker.rosco.api.BakeStatus
+import com.netflix.spinnaker.rosco.endpoint.StatusEndpoint
+import com.netflix.spinnaker.rosco.persistence.RedisBackedBakeStore
+import spock.lang.Specification
+import spock.lang.Subject
+
+class StatusEndpointSpec extends Specification {
+
+  private static final String JOB_ID = "123"
+  private static runningBakeStatus = new BakeStatus(id: JOB_ID,
```
Comment (duftler) [inline rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy]: Please align constructor params.

File: rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+    then:
+    instanceInfo.bakes.isEmpty()
+    instanceInfo.status == "IDLE"
+  }
+
+  void 'redis exception on get incomplete bakes'() {
+    setup:
+    def bakeStoreMock = Mock(RedisBackedBakeStore)
+    bakeStoreMock.getThisInstanceIncompleteBakeIds() >> {throw new RuntimeException()}
```
Comment (duftler) [inline rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy]: Please surround `throw new RuntimeException()` with spaces.

File: rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    then:
+    thrown(RuntimeException)
+  }
+
+  void 'redis exception on get bakes result'() {
+    setup:
+    def bakeStoreMock = Mock(RedisBackedBakeStore)
+    bakeStoreMock.getThisInstanceIncompleteBakeIds() >> Sets.newHashSet(JOB_ID)
+    bakeStoreMock.retrieveBakeStatusById(JOB_ID) >> {throw new RuntimeException()}
```
Comment (duftler) [inline rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy]: Please surround `throw new RuntimeException()` with spaces.

File: rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    @Subject
+    def statusEndpoint = new StatusEndpoint(bakeStoreMock)
+
+    when:
+    statusEndpoint.invoke()
+
+    then:
+    thrown(RuntimeException)
+  }
+
```
Comment (duftler) [inline rosco-web/src/test/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpointSpec.groovy]: Extraneous blank lines.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  private List<BakeStatus> bakesStatus(incompleteBakes) {
+    def status = []
+    for (bakeId in incompleteBakes) {
+      def bakeStatus = bakeStore.retrieveBakeStatusById(bakeId)
+      if (bakeStatus) {
+        status << bakeStatus
+      }
+    }
+    status
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy]: Please use `return` operator.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/mvc/StatusMvcEndpoint.groovy [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  @RequestMapping(method = RequestMethod.GET)
+  @ResponseBody
+  @Override
+  public Object invoke() {
+    if (!getDelegate().isEnabled()) {
+      return new ResponseEntity<Map<String, String>>(Collections.singletonMap(
+          "message", "This endpoint is disabled"), HttpStatus.NOT_FOUND)
+    }
+    super.invoke()
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/mvc/StatusMvcEndpoint.groovy]: Please use `return` operator.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy [hunk 9]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy]: Please name the package `com.netflix.spinnaker.rosco.endpoints` to be consistent with its sibling `com.netflix.spinnaker.rosco.controllers` package.

**Comment by duftler** [comment]
This is really nice. Just a few minor formatting comments.

**Comment by duftler** [comment]
Note that the sample configuration is off a bit. Instead of this:
```
endpoint.status.enable = true
endpoint.status.sensitive = false
```
It should be:
```
endpoints.status.enabled: true
endpoints.status.sensitive: false
```

**Comment by duftler** [comment]
I would enable this by default, and set `sensitive` to `false` by default. I can't see the connection between the `sensitive` setting and auth, but I'm fairly ignorant of the boot particulars. In any case, this endpoint isn't any more sensitive than the others that are already exposed.

**Comment by duftler** [comment]
Last suggestion (more of a question really): Any reason to not expose all in-flight bakes? That is, rely on `bakeStore.getAllIncompleteBakeIds()` instead of `bakeStore.getThisInstanceIncompleteBakeIds()`? Could always add in the `roscoInstanceId` of the instance that handled the request.

**Comment by duftler** [comment]
Please also add this to the list of swagger patterns: https://github.com/spinnaker/rosco/blob/master/rosco-web/config/rosco.yml#L273

**Comment by duftler** [comment]
I will be using this constantly. Already relying on your branch locally...

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy (hunk 9 above)
Comment (tgracchus) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy]: ok

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy (hunk 1 above)
Comment (tgracchus) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy]: ok

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy (hunk 7 above)
Comment (tgracchus) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/endpoint/StatusEndpoint.groovy]: ok

**Comment by duftler** [comment]
I would not be opposed to exposing both endpoints.

**Pull request merged**